Seems incompatible with Keycloak 21

## Description
After having migrated to Keycloak 21 (Quarkus) I get "Internal Server Error" whenever I try to login. The logs show messages like these:

`ERROR [io.quarkus.vertx.http.runtime.QuarkusErrorHandler] (executor-thread-7) HTTP Request to /auth/realms/jake/metrics failed, error id: fdebc248-5921-4c51-8db2-2bcbcff3607e-57: org.jboss.resteasy.spi.UnhandledException: java.lang.IllegalArgumentException: Failed to register Collector of type Counter: keycloak_login_attempts_total is already in use by another Collector of type Counter`

Keycloak 21 changed metrics implementation from SmallRye to Micrometer, maybe this clashes with this project somehow?

## Steps to reproduce
1. Install and enable keycloak-metrics-spi v2.5.3 on Keycloak 20
2. Migrate to Keycloak 21
3. Attempt to login to the admin console

I seem to have the same problem but already have issues when starting Keycloak.
This is what i get from the logs:

2023-03-02 10:31:22,776 DEBUG [io.quarkus.micrometer.runtime.MicrometerRecorder] (main) Discovered global MeterFilters : [io.micrometer.core.instrument.config.MeterFilter$9@7573c7b5, io.micrometer.core.instrument.config.MeterFilter$9@73da7f08, io.micrometer.core.instrument.config.MeterFilter$9@717a8a76, io.micrometer.core.instrument.config.MeterFilter$9@62de73eb]
2023-03-02 10:31:22,776 DEBUG [io.quarkus.micrometer.runtime.MicrometerRecorder] (main) Configuring Micrometer registries : [class io.micrometer.prometheus.PrometheusMeterRegistry]
2023-03-02 10:31:22,776 DEBUG [io.quarkus.micrometer.runtime.MicrometerRecorder] (main) Discovered MeterFilters for class io.micrometer.prometheus.PrometheusMeterRegistry: []
2023-03-02 10:31:22,792 DEBUG [io.quarkus.micrometer.runtime.MicrometerRecorder] (main) Adding configured registry class io.micrometer.prometheus.PrometheusMeterRegistry
2023-03-02 10:31:23,385 ERROR [org.infinispan.CONFIG] (keycloak-cache-init) ISPN000660: DefaultCacheManager start failed, stopping any running components: org.infinispan.commons.CacheConfigurationException: ISPN000085: Error while trying to create a channel using the specified configuration file: default-configs/default-jgroups-udp.xml
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.buildChannel(JGroupsTransport.java:758)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.initChannel(JGroupsTransport.java:504)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.start(JGroupsTransport.java:485)
	at org.infinispan.remoting.transport.jgroups.CorePackageImpl$1.start(CorePackageImpl.java:42)
	at org.infinispan.remoting.transport.jgroups.CorePackageImpl$1.start(CorePackageImpl.java:27)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.invokeStart(BasicComponentRegistryImpl.java:617)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:608)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:577)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:808)
	at org.infinispan.metrics.impl.MetricsCollector.start(MetricsCollector.java:78)
	at org.infinispan.metrics.impl.CorePackageImpl$1.start(CorePackageImpl.java:41)
	at org.infinispan.metrics.impl.CorePackageImpl$1.start(CorePackageImpl.java:34)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.invokeStart(BasicComponentRegistryImpl.java:617)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:608)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:577)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:808)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startDependencies(BasicComponentRegistryImpl.java:635)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:599)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:577)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:808)
	at org.infinispan.factories.AbstractComponentRegistry.internalStart(AbstractComponentRegistry.java:357)
	at org.infinispan.factories.AbstractComponentRegistry.start(AbstractComponentRegistry.java:250)
	at org.infinispan.manager.DefaultCacheManager.internalStart(DefaultCacheManager.java:774)
	at org.infinispan.manager.DefaultCacheManager.start(DefaultCacheManager.java:742)
	at org.infinispan.manager.DefaultCacheManager.<init>(DefaultCacheManager.java:406)
	at org.keycloak.quarkus.runtime.storage.legacy.infinispan.CacheManagerFactory.startCacheManager(CacheManagerFactory.java:96)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1589)
Caused by: java.lang.IllegalArgumentException: failed initializing component ThreadPool in protocol UDP: java.lang.IllegalStateException: failed to create virtual thread pool: java.lang.UnsupportedOperationException: Preview Features not enabled, need to run with --enable-preview
	at org.jgroups.stack.ProtocolStack.lambda$initComponents$7(ProtocolStack.java:861)
	at org.jgroups.util.Util.forAllComponents(Util.java:3368)
	at org.jgroups.stack.ProtocolStack.initComponents(ProtocolStack.java:839)
	at org.jgroups.stack.ProtocolStack.initProtocolStack(ProtocolStack.java:795)
	at org.jgroups.stack.ProtocolStack.setup(ProtocolStack.java:438)
	at org.jgroups.JChannel.init(JChannel.java:899)
	at org.jgroups.JChannel.<init>(JChannel.java:123)
	at org.jgroups.JChannel.<init>(JChannel.java:113)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.buildChannel(JGroupsTransport.java:756)
	... 29 more

Hopefully this helps in finding what is wrong.

--- a/build.gradle
+++ b/build.gradle
@@ -13,11 +13,11 @@ configurations {
 }
 
 group 'org.jboss.aerogear'
-version '2.5.4-SNAPSHOT'
+version '3.0.0-SNAPSHOT'
 
 apply plugin: 'java'
 
-sourceCompatibility = 1.8
+sourceCompatibility = 17
 
 repositories {
     mavenCentral()

--- a/gradle.properties
+++ b/gradle.properties
@@ -1,2 +1,2 @@
-keycloakVersion=15.0.2
-prometheusVersion=0.9.0
+keycloakVersion=21.0.1
+prometheusVersion=0.16.0

--- a/pom.xml
+++ b/pom.xml
@@ -7,15 +7,15 @@
     <artifactId>keycloak-metrics-spi</artifactId>
     <groupId>org.jboss.aerogear</groupId>
     <packaging>jar</packaging>
-    <version>2.5.3-SNAPSHOT</version>
+    <version>3.0.0-SNAPSHOT</version>
 
     <properties>
-        <java.version>11</java.version>
-        <keycloak.version>15.0.2</keycloak.version>
-        <prometheus.version>0.9.0</prometheus.version>
+        <java.version>17</java.version>
+        <keycloak.version>21.0.1</keycloak.version>
+        <prometheus.version>0.16.0</prometheus.version>
         <version.compiler.maven.plugin>3.5.1</version.compiler.maven.plugin>
-        <maven.compiler.source>11</maven.compiler.source>
-        <maven.compiler.target>11</maven.compiler.target>
+        <maven.compiler.source>17</maven.compiler.source>
+        <maven.compiler.target>17</maven.compiler.target>
     </properties>
 
     <dependencies>

--- a/src/main/java/org/jboss/aerogear/keycloak/metrics/PrometheusExporter.java
+++ b/src/main/java/org/jboss/aerogear/keycloak/metrics/PrometheusExporter.java
@@ -152,7 +152,7 @@ public final class PrometheusExporter {
         final boolean URI_METRICS_ENABLED = Boolean.parseBoolean(System.getenv("URI_METRICS_ENABLED"));
         if (URI_METRICS_ENABLED){
             responseTotal = Counter.build()
-            .name("keycloak_response_total")
+            .name("keycloak_response")
             .help("Total number of responses")
             .labelNames("code", "method", "resource", "uri")
             .register();
@@ -171,7 +171,7 @@ public final class PrometheusExporter {
             .register();
         } else {
             responseTotal = Counter.build()
-            .name("keycloak_response_total")
+            .name("keycloak_response")
             .help("Total number of responses")
             .labelNames("code", "method", "resource")
             .register();

--- a/src/test/java/org/jboss/aerogear/keycloak/metrics/PrometheusExporterTest.java
+++ b/src/test/java/org/jboss/aerogear/keycloak/metrics/PrometheusExporterTest.java
@@ -288,7 +288,7 @@ public class PrometheusExporterTest {
     public void shouldCorrectlyRecordResponseErrors() throws IOException {
         environmentVariables.set("URI_METRICS_ENABLED", "true");
         PrometheusExporter.instance().recordResponseError(500, "POST", "admin,admin/serverinfo", "auth/realm");
-        assertGenericMetric("keycloak_response_errors", 1,
+        assertGenericMetric("keycloak_response_errors_total", 1,
             tuple("code", "500"), tuple("method", "POST"), tuple("resource", "admin,admin/serverinfo"), tuple("uri", "auth/realm"));
     }
 
@@ -349,7 +349,7 @@ public class PrometheusExporterTest {
 
             final StringBuilder builder = new StringBuilder();
 
-            builder.append(metricName).append("{");
+            builder.append(metricName).append("_total").append("{");
             builder.append("realm").append("=\"").append(realm).append("\",");
 
             for (Tuple<String, String> label : labels) {

